apps/publisher/src/app.tsx

```diff
@@ -62,220 +65,277 @@ function App() {
     };
   }, []);
 
-  const { showError } = useNotification();
+  const {
+    addStream,
+    applyConstraints,
+    cameraList,
+    initDefaultStream,
+    microphoneList,
+    // TODO: remove and reset streams
+    // removeStream,
+    // reset,
+    streams,
+    // TODO: per-stream audio/video toggling
+    // toggleAudio,
+    // toggleVideo,
+  } = useMultiMediaDevices();
 
   const {
+    bitrateList,
+    codecList,
+    shareUrl,
+    sources,
     startStreamingSource,
     stopStreamingSource,
-    updateSourceMediaStream,
     updateSourceBitrate,
-    sources,
-    codecList,
-    bitrateList,
+    updateSourceMediaStream,
     viewerCount,
-    shareUrl,
   } = usePublisher({
```

Comment:
As I mentioned above, is it better to put the publisher-related states of each source in this hook?

Reply:
We will solve this in issue #241